### PR TITLE
status hook for the container network

### PR DESCRIPTION
**Pull request by rajatchopra**
Extending the network plugin to include a status hook that gets the Pod's IPAddress information as decided by the network plugin. Until now, even though the network could be plumbed by a network plugin, the IPAddress of the pods was obtained from container runtime (e.g. docker inspect) only.

A 'status' API (and a corresponding hook in the exec plugin) lets the plugin give out the IPAddress as it deems fit. This feature is particularly necessary if the network plugin chose its own IPAM and not the one provided by container-runtime.

Marking this as a Work-In-Progress, feedback please.

@errordeveloper @jainvipin @mapuri @thockin @mrunalp @joeswaminathan

### CODE REVIEW

File: pkg/kubelet/network/exec/exec.go [hunk 1]
```diff
@@ -136,3 +141,15 @@ func (plugin *execNetworkPlugin) TearDownPod(namespace string, name string, id k
 	glog.V(5).Infof("TearDownPod 'exec' network plugin output: %s, %v", string(out), err)
 	return err
 }
+
+func (plugin *execNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.NetworkStatus, error) {
+	out, err := utilexec.New().Command(plugin.getExecutable(), statusCmd, namespace, name, string(id)).CombinedOutput()
+	glog.V(5).Infof("Status 'exec' network plugin output: %s, %v", string(out), err)
+	fmt.Printf("Status 'exec' network plugin output: %s, %v", string(out), err)
```
Comment (jainvipin) [inline pkg/kubelet/network/exec/exec.go]: minor: do we need an extra Printf?

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -280,8 +280,15 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
 		result.status.State.Running = &api.ContainerStateRunning{
 			StartedAt: util.NewTime(inspectResult.State.StartedAt),
 		}
-		if containerName == PodInfraContainerName && inspectResult.NetworkSettings != nil {
-			result.ip = inspectResult.NetworkSettings.IPAddress
+		if dm.networkPlugin.Name() != network.DefaultPluginName && containerName == PodInfraContainerName {
+			netStatus, err := dm.networkPlugin.Status(pod.Namespace, pod.Name, kubeletTypes.DockerID(dockerID))
+			if err == nil {
+				result.ip = netStatus.IPv4.String()
```
Comment (jainvipin) [inline pkg/kubelet/dockertools/manager.go]: `NetworkStatus` can be a rich and growing set of useful information. Perhaps keeping it as structured data that is backward compatible would be the right way to go. However consider following scenarios, if we want to change anything to the `NetworkStatus`:
- If a pod is allocated only a v6 address, the above code/logic wouldn't work, i.e. we may need to specify if v4 and/or v6 address is valid
- A pod could be assigned multiple interfaces, and thus multiple IP addresses, juxtaposing all that as `result.ip` may work but I am not sure if it is desired
- If driver wants to display other information like policy, etc.

File: pkg/kubelet/network/exec/exec.go (hunk 1 above)
Comment (rajatchopra) [inline pkg/kubelet/network/exec/exec.go]: My bad. Leftover from my local testing. Will get it on the next update.

File: pkg/kubelet/dockertools/manager.go (hunk 2 above)
Comment (rajatchopra) [inline pkg/kubelet/dockertools/manager.go]: Valid points. Multiple interfaces and displaying other policy information would mean enhancing the 'result' structure which I believe is out of scope of this PR, though.
'result.ip' is what services look at today the endpoint watch does not support multiple addresses.

Comment (jainvipin) [inline pkg/kubelet/dockertools/manager.go]: I wasn't pushing to get those changes in this PR; my point was if NetworkStatus could be organized differently now to favor these later it would be helpful. For this PR I think we should add `ipv4.valid` and `ipv6.valid`, and if possible arrays of v4 and v6 addresses.

Comment (mrunalp) [inline pkg/kubelet/dockertools/manager.go]: Agree that it makes sense to make it extensible.
Does something like this work?
```go

type struct InterfaceInfo {
    Iface net.Interface
    IPAddrs []net.IP
}

type struct NetworkStatus {
    IPv4 net.IP  // Primary ipv4 address of the pod
    IPv6 net.IP  // Primary ipv6 address of the pod
    NetworkInterfaces []net.InterfaceInfo
}
```

Comment (jainvipin) [inline pkg/kubelet/dockertools/manager.go]: @rajatchopra, @mrunalp - yes, works for me

one minor follow-up: how do we identify if IPv4 or IPv6 is valid? `result.ip` could be filled in with a valid address (for the case I was alluding for ipv6 deployment).

Comment (rajatchopra) [inline pkg/kubelet/dockertools/manager.go]: How about just storing the list of InterfaceInfo, and we always pick the first one. The plugin can store IPv4 or IPv6 as it pleases as golang's net.IP will store either.
```go
type InterfaceInfo struct {
    Iface net.Interface
    IPAddrs []net.IP
}

type NetworkStatus struct {
     NetworkInterfaces []net.InterfaceInfo
}

// kube code.. null checks excluded
result.ip = netStatus.NetworkInterfaces[0].IPAddrs[0].String()
```

Alternately, we can store a primary IPAddress at the top (like docker's NetworkSettings.IPAddress):
```go
type InterfaceInfo struct {
    Iface net.Interface
    IPAddrs []net.IP
}

type NetworkStatus struct {
    IPAddress net.IP  // Primary ipv4/ipv6 address of the pod
    NetworkInterfaces []net.InterfaceInfo
}
```

Comment (joeswaminathan) [inline pkg/kubelet/dockertools/manager.go]: Can primary be an attribute in the Interface Info rather than repeating it in NetworkStatus ?
What about routes ?  The container network plugins are allowing the routes to be published. May be it could be attempted later ?

In case of DHCP, if the IP address changes what is the mechanism to push the new IP to the Master ?

Comment (rajatchopra) [inline pkg/kubelet/dockertools/manager.go]: @joeswaminathan Putting primary inside InterfaceInfo will result in indecision about which interface is primary. Yes, we can add routes or other policy metadata later (note that kube does not consume them yet).
Regarding DHCP, the status hook is called on periodically, so any changes will be captured too.

**Comment by rajatchopra** [comment]
@jainvipin Replaced IPv4/v6 with just one field at the top level struct that represents the primary IPAddress that kube should store for the pod (docker inspect style).

**Comment by jainvipin** [comment]
@rajatchopra - LGTM!

Thanks for the important contribution. We'll use it.

**Comment by rajatchopra** [comment]
@thockin I think this one is ready. Also, is there anything I have to do for my CLA under RedHat?

**Comment by saad-ali** [comment]
Assigned @thockin 
v1.0 triage needed

**Comment by FlorianOtel** [comment]
@thockin as per today's discussion at DockerCon: please prioritise this: from an networking plugin pov we heed this status hook for checking IP address allocation / @rajatchopra @abhat @mrunalp

Kindest thanks yet again.

File: pkg/kubelet/dockertools/manager.go [hunk 3]
```diff
@@ -288,8 +288,15 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
 		result.status.State.Running = &api.ContainerStateRunning{
 			StartedAt: util.NewTime(inspectResult.State.StartedAt),
 		}
-		if containerName == PodInfraContainerName && inspectResult.NetworkSettings != nil {
-			result.ip = inspectResult.NetworkSettings.IPAddress
+		if dm.networkPlugin.Name() != network.DefaultPluginName && containerName == PodInfraContainerName {
```
Comment (marun) [inline pkg/kubelet/dockertools/manager.go]: Consider simplifying as follows:

if containerName == PodInfraContainerName {
  if dm.networkPlugin.Name != network.DefaultPluginName {
  ...
  } else if inspectResults.NetworkSettings != nil {
  ...
  }
}

File: pkg/kubelet/dockertools/manager.go [hunk 4]
```diff
@@ -288,8 +288,15 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
...
+			netStatus, err := dm.networkPlugin.Status(pod.Namespace, pod.Name, kubeletTypes.DockerID(dockerID))
+			if err == nil {
```
Comment (marun) [inline pkg/kubelet/dockertools/manager.go]: I assume it's safe for result.ip to not be populated, but wouldn't it be desirable to at least log any error in the unmarshalling performed in Status()?

File: pkg/kubelet/network/plugins.go [hunk 5]
```diff
@@ -47,6 +48,20 @@ type NetworkPlugin interface {
 
 	// TearDownPod is the method called before a pod's infra container will be deleted
 	TearDownPod(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) error
+
+	// Status is the method called to obtain the ipv4 and pv6 addresses of the container
```
Comment (marun) [inline pkg/kubelet/network/plugins.go]: pv6 -> ipv6

**Comment by errordeveloper** [comment]
@thockin what's the status of this from your pov?

**Comment by thockin** [comment]
Sorry for latency - I've had it up on my screen for 2 weeks.  Reviewing now.

File: pkg/kubelet/dockertools/manager.go (hunk 3 above)
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: I like this proposed layout better but it's a minor point - if no other changes are necessary I won't require it :)

File: pkg/kubelet/dockertools/manager.go (hunk 4 above)
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: +1 - if the plugin returns an error that's an unexpected thing, right?  Not something to hide..

File: pkg/kubelet/network/exec/exec.go [hunk 6]
```diff
@@ -29,6 +29,9 @@ limitations under the License.
 //         - setup, called after the infra container of a pod is
 //                created, but before other containers of the pod are created
 //         - teardown, called before the pod infra container is killed
+//         - status, called at regular intervals and is supposed to return a json
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: Design: Why is this a periodic pull rather than something returned from SetUpPod() ?  Is there the expectation that a pod can change IPs after creation?  That might break a number of other components like the Service->Endpoint generation loop (not sure).

File: pkg/kubelet/network/exec/exec.go [hunk 7]
```diff
@@ -29,6 +29,9 @@ limitations under the License.
...
+//                formatted output indicating the pod's IPAddresses
+//                e.g. { "IPAddress" : "10.20.30.40" }
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: If we're passing around JSON blobs, we probably need to version them, at least.  I don't know if we have any other examples of this in the system yet, so I want to be careful to set good precedent.

Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: This JSON example does not seem to match the struct defined below.

File: pkg/kubelet/network/plugins.go [hunk 8]
```diff
@@ -47,6 +48,20 @@ type NetworkPlugin interface {
...
+}
+
+type InterfaceInfo struct {
+	Iface   net.Interface
+	IPAddrs []net.IP
+}
+
+// Stores the network status of a pod (currently just the network addresses and interfaces)
+type NetworkStatus struct {
+	IPAddress         net.IP // Primary ipv4/ipv6 address of the pod
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: First: this struct needs an apiVersion field.

Comment (thockin) [inline pkg/kubelet/network/plugins.go]: What does it mean to have a "primary" IP, when you also allow for an arbitrary number of interfaces each of which can have an arbitrary number of IPs?

Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Which fields are required vs optional?

Comment (thockin) [inline pkg/kubelet/network/plugins.go]: any struct which intended to be JSON'ed probably should have explicit struct tags indicating JSON name and omitempty (if applicable) for optional fields.

File: pkg/kubelet/dockertools/manager.go [hunk 9]
```diff
@@ -288,8 +288,15 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
...
+				result.ip = netStatus.IPAddress.String()
```
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: noop plugin returns (nil, nil) - that is not handled here.

File: pkg/kubelet/network/plugins.go [hunk 10]
```diff
@@ -120,3 +135,7 @@ func (plugin *noopNetworkPlugin) SetUpPod(namespace string, name string, id kube
 func (plugin *noopNetworkPlugin) TearDownPod(namespace string, name string, id kubeletTypes.DockerID) error {
 	return nil
 }
+
+func (plugin *noopNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*NetworkStatus, error) {
+	return nil, nil
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: This does not seem to handled by kubelet - I expect a nil dereference.  It should either handle a nil return and fall back on docker inspect or this should be a required (and tested) non-nil return.

**Comment by rajatchopra** [comment]
@thockin Thanks for taking a look here. Updated the PR. Handled the error messages, nil/nil case and json explicit strings for the struct. Also docker's status will be overridden only if the network plugin is valid and returns a valid IP.

Still working on a suitable way to include the api version for the json blob.

File: pkg/kubelet/network/exec/exec.go (hunk 6 above)
Comment (rajatchopra) [inline pkg/kubelet/network/exec/exec.go]: At this point we are just following the model of how container runtime is called and then IP address is obtained from the regular Status calls.

I am expecting that at some point this Status call will include policy structures too, or custom info about the network configuration of the pod (see #12170)

File: pkg/kubelet/dockertools/manager.go [hunk 11]
```diff
@@ -288,8 +288,19 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
 		result.status.State.Running = &api.ContainerStateRunning{
 			StartedAt: util.NewTime(inspectResult.State.StartedAt),
 		}
-		if containerName == PodInfraContainerName && inspectResult.NetworkSettings != nil {
-			result.ip = inspectResult.NetworkSettings.IPAddress
+		if containerName == PodInfraContainerName {
+			if containerName == PodInfraContainerName && inspectResult.NetworkSettings != nil {
```
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: first clause of this conditional is redundant

File: pkg/kubelet/network/exec/exec.go (hunk 6 above)
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: I'm OK with it, just wanted to understand.  I'm a little concerned about overhead, but let's just keep an eye on it.

File: pkg/kubelet/network/exec/exec.go [hunk 12]
```diff
@@ -70,6 +74,8 @@ const (
 	initCmd     = "init"
 	setUpCmd    = "setup"
 	tearDownCmd = "teardown"
+	statusCmd   = "status"
+	execDir     = "/usr/libexec/kubernetes/kubelet-plugins/net/exec/"
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: This line is unrelated to this PR - didn't we make this a flag recently?

File: pkg/kubelet/network/plugins.go (hunk 5 above)
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: still needs to be fixed

File: pkg/kubelet/network/plugins.go [hunk 13]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
 
 	// TearDownPod is the method called before a pod's infra container will be deleted
 	TearDownPod(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) error
+
+	// Status is the method called to obtain the ipv4 and pv6 addresses of the container
+	Status(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) (*NetworkStatus, error)
+}
+
+type InterfaceInfo struct {
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Needs a godoc-style comment

File: pkg/kubelet/network/plugins.go [hunk 14]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+	Iface   net.Interface `json:"iFace" description:"A network interface on the pod"`
+	IPAddrs []net.IP      `json:"ipAddrs" description:"List of IP Addresses associated with the given interface"`
+}
+
+// Stores the network status of a pod (currently just the network addresses and interfaces)
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: comment should be godoc style

File: pkg/kubelet/network/plugins.go (hunk 5 above)
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: comment on return values, sepcifically that a nil NetworkStatus return means "use docker's info"

File: pkg/kubelet/network/plugins.go [hunk 15]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
 
 	// TearDownPod is the method called before a pod's infra container will be deleted
 	TearDownPod(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) error
+
+	// Status is the method called to obtain the ipv4 and pv6 addresses of the container
+	Status(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) (*NetworkStatus, error)
+}
+
+type InterfaceInfo struct {
+	Iface   net.Interface `json:"iFace" description:"A network interface on the pod"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Don't abbreviate - interface

File: pkg/kubelet/network/plugins.go [hunk 16]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+	IPAddrs []net.IP      `json:"ipAddrs" description:"List of IP Addresses associated with the given interface"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: don't appreviate - either IPs or IPAddresses

File: pkg/kubelet/network/plugins.go [hunk 17]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+}
+
+// Stores the network status of a pod (currently just the network addresses and interfaces)
+type NetworkStatus struct {
+	// Primary ipv4/ipv6 address of the pod
+	IPAddress net.IP `json:"ipAddress" description:"Primary address of the pod"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: I still don't understand what it means to be a "primary" IP vs all of the other IPs that could be present in this structure.

File: pkg/kubelet/network/plugins.go [hunk 18]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+	Status(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) (*NetworkStatus, error)
+}
+
+type InterfaceInfo struct {
+	Iface   net.Interface `json:"iFace" description:"A network interface on the pod"`
+	IPAddrs []net.IP      `json:"ipAddrs" description:"List of IP Addresses associated with the given interface"`
+}
+
+// Stores the network status of a pod (currently just the network addresses and interfaces)
+type NetworkStatus struct {
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: as a JSON struct this needs a Kind and APIVersion field, just like api.TypeMeta.  In fact, just embed an api.TypeMeta.  See plugin/pkg/scheduler/api/types.go for precedent.

File: pkg/kubelet/network/exec/exec.go [hunk 19]
```diff
@@ -131,3 +137,17 @@ func (plugin *execNetworkPlugin) TearDownPod(namespace string, name string, id k
...
+func (plugin *execNetworkPlugin) Status(namespace string, name string, id kubeletTypes.DockerID) (*network.NetworkStatus, error) {
+	out, err := utilexec.New().Command(plugin.getExecutable(), statusCmd, namespace, name, string(id)).CombinedOutput()
+	glog.V(5).Infof("Status 'exec' network plugin output: %s, %v", string(out), err)
+	if err != nil {
+		return nil, err
+	}
+	if string(out) == "" {
+		return nil, nil
+	}
+	networkStatus := &network.NetworkStatus{}
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: check the Kind and APIVersion before unmarshalling.

File: pkg/kubelet/network/plugins.go (hunk 18 above)
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Also update example JSON

File: pkg/kubelet/network/plugins.go [hunk 20]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+	// Primary ipv4/ipv6 address of the pod
+	IPAddress net.IP `json:"ipAddress" description:"Primary address of the pod"`
+	// List of network interfaces that may be optionally provided
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Comment on what these would be used for, if provided.  If the answer is "nothing", do we really need this?

File: pkg/kubelet/network/plugins.go [hunk 21]
```diff
@@ -47,6 +48,22 @@ type NetworkPlugin interface {
...
+}
+
+type InterfaceInfo struct {
+	Iface   net.Interface `json:"iFace" description:"A network interface on the pod"`
+	IPAddrs []net.IP      `json:"ipAddrs" description:"List of IP Addresses associated with the given interface"`
+}
+
+// Stores the network status of a pod (currently just the network addresses and interfaces)
+type NetworkStatus struct {
+	// Primary ipv4/ipv6 address of the pod
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Comment on the fact that this will supercede the docker-reported IP.

File: pkg/kubelet/network/exec/exec_test.go [hunk 22]
```diff
@@ -51,7 +51,7 @@ func installPluginUnderTest(t *testing.T, vendorName string, plugName string) {
 	if err != nil {
 		t.Errorf("Failed to set exec perms on plugin")
 	}
-	writeStr := fmt.Sprintf("#!/bin/bash\necho -n $@ &> %s", path.Join(pluginDir, plugName+".out"))
+	writeStr := fmt.Sprintf("#!/bin/bash\nif [ \"$1\" == \"status\" ]; then echo -n '{ \"ipaddress\" : \"10.20.30.40\" }'; fi\n echo -n $@ &> %s", path.Join(pluginDir, plugName+".out"))
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec_test.go]: can you turn this into a proper multi-line backtick-enclosed string?  it's complex enough now

File: pkg/kubelet/network/exec/exec_test.go [hunk 23]
```diff
@@ -177,3 +177,28 @@ func TestPluginTearDownHook(t *testing.T) {
...
 }
+
+func TestPluginStatusHook(t *testing.T) {
+	pluginName := fmt.Sprintf("test%d", rand.Intn(1000))
+	defer tearDownPlugin(pluginName)
+	installPluginUnderTest(t, "", pluginName)
+
+	plug, err := network.InitNetworkPlugin(ProbeNetworkPlugins(testPluginPath), pluginName, network.NewFakeHost(nil))
+
+	ip, err := plug.Status("namespace", "name", "dockerid2345")
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec_test.go]: We really ought to test the plugin returning "" as well as error cases.  This just got a lot more complicated to test

File: pkg/kubelet/network/plugins.go (hunk 20 above)
Comment (rajatchopra) [inline pkg/kubelet/network/plugins.go]: If you track the discussion early in the PR, there have been suggestions from the community to extend the information returned. Today, the PodStatus only stores one IP address, but likely that network plugins will have multiple addresses and interfaces (libnetwork for example). Also helpful for the case for IPv6.

So, while the 'exec' plugin flow does not really use the extra bits, there is scope for a plugin to want to manage more. Maybe through annotations.

I would invite @jainvipin to comment about this.

**Comment by rajatchopra** [comment]
Feedback incorporated with updated PR.
One thing remains:
The fallback testcase - when plugin returns "", docker's info will be used and no test covers that yet. Will think about an e2e for this, or possibly use some of the fake docker bits.

File: pkg/kubelet/network/exec/exec.go [hunk 24]
```diff
@@ -29,6 +29,31 @@ limitations under the License.
 //         - setup, called after the infra container of a pod is
 //                created, but before other containers of the pod are created
 //         - teardown, called before the pod infra container is killed
+//         - status, called at regular intervals and is supposed to return a json
+//                formatted output indicating the pod's IPAddresses. An empty string value or an erroneous output
+//                will mean the container runtime (docker) will be asked for the PodIP
+//                e.g. {
+//                         "apiVersion" : "latest",
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: let's start with v1beta1 and we can sync it to the main API when we mark this non-alpha

File: pkg/kubelet/network/exec/exec.go [hunk 25]
```diff
@@ -131,3 +159,39 @@ func (plugin *execNetworkPlugin) TearDownPod(namespace string, name string, id k
...
+	if findVersion.Kind != "" && findVersion.Kind != "PodNetworkStatus" {
+		errStr := fmt.Sprintf("Invalid 'kind' returned in network status for pod '%s'. Valid value is 'PodNetworkStatus', got '%s'.", name, findVersion.Kind)
+		return nil, errors.New(errStr)
+	}
+	switch findVersion.APIVersion {
+	case "":
+		fallthrough
+	case "v1":
+		fallthrough
+	case "latest":
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: what's the intent with "latest" ?  We don't do this for the main API, so let's keep this as simple as possible.

File: pkg/kubelet/network/exec/exec.go [hunk 26]
```diff
@@ -131,3 +159,39 @@ func (plugin *execNetworkPlugin) TearDownPod(namespace string, name string, id k
...
+	if err != nil {
+		return nil, err
+	}
+	if string(out) == "" {
+		return nil, nil
+	}
+	findVersion := struct {
+		api.TypeMeta `json:",inline"`
+	}{}
+	err = json.Unmarshal(out, &findVersion)
```
Comment (thockin) [inline pkg/kubelet/network/exec/exec.go]: This is OK for now, but we should really unmarshal into a TypeMeta, check the version and kind, and then unmarshal into a versioned object.  We can do that when we have a second version :)

File: pkg/kubelet/network/plugins.go [hunk 27]
```diff
@@ -47,6 +48,30 @@ type NetworkPlugin interface {
...
+	api.TypeMeta `json:",inline"`
+
+	// IPAddress is the primary ipv4/ipv6 address of the pod. Among other things it is the address that -
+	//   - kube expects to be reachable across the cluster
+	//   - service endpoints are constructed with
+	//   - will be reported in the PodStatus.PodIP field (will override the IP reported by docker)
+	IPAddress net.IP `json:"ipAddress" description:"Primary address of the pod"`
+
+	// NetworkInterfaces is a list of network interfaces that may be optionally provided
+	NetworkInterfaces []InterfaceInfo `json:"networkInterfaces,omitempty"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: I hear the desire to carry more information, but here's my fear.  Things like this that get added but don't have anyone who uses them tend to rot and smell bad.  @jainvipin do we have an immediate use for these?  Maybe we can find a way to expose this info somewhere (it's going to be fun with "experimental" exposures)?  If we kicked this out of *this* PR and commited it in a PR that actually consumes it, would it be a problem?

Comment (thockin) [inline pkg/kubelet/network/plugins.go]: Having a consumer will help vet the optional-ness of this, too.

File: pkg/kubelet/network/plugins.go [hunk 28]
```diff
@@ -47,6 +48,30 @@ type NetworkPlugin interface {
...
+// PodNetworkStatus stores the network status of a pod (currently just the network addresses and interfaces)
+// This struct represents version "v1"
+type PodNetworkStatus struct {
+	api.TypeMeta `json:",inline"`
+
+	// IPAddress is the primary ipv4/ipv6 address of the pod. Among other things it is the address that -
+	//   - kube expects to be reachable across the cluster
+	//   - service endpoints are constructed with
+	//   - will be reported in the PodStatus.PodIP field (will override the IP reported by docker)
+	IPAddress net.IP `json:"ipAddress" description:"Primary address of the pod"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: can we just call this IP - more consistent with the other API objects.  sorry :)

File: pkg/kubelet/network/plugins.go [hunk 29]
```diff
@@ -47,6 +48,30 @@ type NetworkPlugin interface {
...
 	TearDownPod(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) error
+
+	// Status is the method called to obtain the ipv4 and ipv6 addresses of the container
+	Status(namespace string, name string, podInfraContainerID kubeletTypes.DockerID) (*PodNetworkStatus, error)
+}
+
+// InterfaceInfo is a structure to store network interface of a pod along with its associated IP addresses
+type InterfaceInfo struct {
+	NetworkInterface net.Interface `json:"networkInterface" description:"A network interface on the pod, e.g. eth1(index:1,mtu:1500,name:eth1..)"`
+	IPAddresses      []net.IP      `json:"ipAddresses" description:"List of IP Addresses associated with the given interface"`
```
Comment (thockin) [inline pkg/kubelet/network/plugins.go]: IPs

**Comment by thockin** [comment]
Mostly small points left

File: pkg/kubelet/network/plugins.go (hunk 27 above)
Comment (rajatchopra) [inline pkg/kubelet/network/plugins.go]: @thockin I would agree with you here. The PR that uses it should be the PR that introduces it. Chops on, update coming.

**Comment by rajatchopra** [comment]
@thockin hopefully the final review from you. Thanks.

File: pkg/kubelet/dockertools/manager.go [hunk 30]
```diff
@@ -288,8 +288,19 @@ func (dm *DockerManager) inspectContainer(dockerID, containerName, tPath string)
...
 		}
-		if containerName == PodInfraContainerName && inspectResult.NetworkSettings != nil {
-			result.ip = inspectResult.NetworkSettings.IPAddress
+		if containerName == PodInfraContainerName {
+			if inspectResult.NetworkSettings != nil {
+				result.ip = inspectResult.NetworkSettings.IPAddress
+			}
+			// override the above if a network plugin exists
+			if dm.networkPlugin.Name() != network.DefaultPluginName {
+				netStatus, err := dm.networkPlugin.Status(pod.Namespace, pod.Name, kubeletTypes.DockerID(dockerID))
```
Comment (thockin) [inline pkg/kubelet/dockertools/manager.go]: What's the status of net plugin support in the rkt path?  Does it even apply?

**Comment by thockin** [comment]
I find nothing else to pick at.

LGTM

**Comment by jszczepkowski** [comment]
@k8s-bot ok to test

File: pkg/kubelet/dockertools/manager.go (hunk 30 above)
Comment (rajatchopra) [inline pkg/kubelet/dockertools/manager.go]: It applies, but other parts of rkt need to be brought up to speed before this can be useful.
